server/src/main/java/org/elasticsearch/cluster/coordination/JoinReasonService.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.service.ClusterApplierService;
+import org.elasticsearch.cluster.service.MasterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.core.Tuple;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.LongSupplier;
+
+import static org.elasticsearch.cluster.coordination.Coordinator.Mode.CANDIDATE;
+
+/**
+ * Tracks nodes that were recently in the cluster, and uses this information to give extra details if these nodes rejoin the cluster.
+ */
+public class JoinReasonService {
+
+    private final LongSupplier relativeTimeInMillisSupplier;
+
+    // only used on cluster applier thread to detect changes, no need for synchronization, and deliberately an Object
+    private Object discoveryNodes;
```

Comment:
Mainly because it doesn't need to be anything more specific - this way we reduce the risk that a future change tries to use its contents without proper synchronization.